staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go

```diff
@@ -59,3 +64,82 @@ type InTreePlugin interface {
 	// RepairVolumeHandle generates a correct volume handle based on node ID information.
 	RepairVolumeHandle(volumeHandle, nodeID string) (string, error)
 }
+
+// getTopology returns the current topology zones with the given key.
+func getTopology(pv *v1.PersistentVolume, key string) []string {
+	if pv.Spec.NodeAffinity == nil ||
+		pv.Spec.NodeAffinity.Required == nil ||
+		len(pv.Spec.NodeAffinity.Required.NodeSelectorTerms) < 1 {
+		return nil
+	}
+	for i := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms {
+		for _, r := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions {
+			if r.Key == key {
+				return r.Values
+			}
+		}
+	}
+	return nil
+}
+
+// recordTopology writes the topology to the given PV.
+// If topology already exists with the given key, assume the content is already accurate.
+func recordTopology(pv *v1.PersistentVolume, key string, zones []string) error {
+	// Make sure there are no duplicate or empty strings
+	filteredZones := sets.String{}
+	for i := range zones {
+		zone := strings.TrimSpace(zones[i])
+		if len(zone) > 0 {
+			filteredZones.Insert(zone)
+		}
+	}
+
+	if filteredZones.Len() < 1 {
+		return fmt.Errorf("there are no valid zones for topology")
+	}
+
+	// Make sure the necessary fields exist
+	if pv.Spec.NodeAffinity == nil {
+		pv.Spec.NodeAffinity = new(v1.VolumeNodeAffinity)
+	}
+
+	if pv.Spec.NodeAffinity.Required == nil {
+		pv.Spec.NodeAffinity.Required = new(v1.NodeSelector)
+	}
+
+	if len(pv.Spec.NodeAffinity.Required.NodeSelectorTerms) == 0 {
+		pv.Spec.NodeAffinity.Required.NodeSelectorTerms = make([]v1.NodeSelectorTerm, 1)
+	}
+
+	// Don't overwrite if topology is already set
+	if len(getTopology(pv, key)) > 0 {
+		return nil
+	}
+
+	pv.Spec.NodeAffinity.Required.NodeSelectorTerms[0].MatchExpressions = append(pv.Spec.NodeAffinity.Required.NodeSelectorTerms[0].MatchExpressions,
```

Comment:
Should we append or overwrite?  I believe appending means that for a PV that already had in-tree NodeAffinity set, then the node needs both the old beta-intree labels + new CSI topology labels

Reply:
I think the conclusion is we should overwrite, not append.  Because eventually nodes are not going to have the beta labels anymore, and appending these is an "AND" operation.

Reply:
Right, I haven't got to this one yet. sweat_smile I'll try to update the code today.

Reply:
Done